core/pkg/artifacts/saver.go

```diff
@@ -99,28 +101,65 @@ func (as *ArtifactSaver) createArtifact() (
 		runId = &as.Artifact.RunId
 	}
 
-	response, err := gql.CreateArtifact(
-		as.Ctx,
-		as.GraphqlClient,
-		as.Artifact.Entity,
-		as.Artifact.Project,
-		as.Artifact.Type,
-		as.Artifact.Name,
-		runId,
-		as.Artifact.Digest,
-		utils.NilIfZero(as.Artifact.Description),
-		aliases,
-		utils.NilIfZero(as.Artifact.Metadata),
-		utils.NilIfZero(as.Artifact.TtlDurationSeconds),
-		utils.NilIfZero(as.HistoryStep),
-		utils.NilIfZero(as.Artifact.DistributedId),
-		as.Artifact.ClientId,
-		as.Artifact.SequenceClientId,
-	)
+	// Older server versions won't support tags, so check the server version to be sure
+	supportsTags, err := as.canSupportArtifactTags()
 	if err != nil {
-		return gql.CreateArtifactCreateArtifactCreateArtifactPayloadArtifact{}, err
+		return gql.CreatedArtifactArtifact{}, err
+	}
+
+	if supportsTags {
+		var tags []gql.TagInput
+		for _, tag := range as.Artifact.Tags {
+			tags = append(tags, gql.TagInput{TagName: tag})
+		}
+
+		response, err := gql.CreateArtifact(
+			as.Ctx,
+			as.GraphqlClient,
+			as.Artifact.Entity,
+			as.Artifact.Project,
+			as.Artifact.Type,
+			as.Artifact.Name,
+			runId,
+			as.Artifact.Digest,
+			utils.NilIfZero(as.Artifact.Description),
+			aliases,
+			tags,
+			utils.NilIfZero(as.Artifact.Metadata),
+			utils.NilIfZero(as.Artifact.TtlDurationSeconds),
+			utils.NilIfZero(as.HistoryStep),
+			utils.NilIfZero(as.Artifact.DistributedId),
+			as.Artifact.ClientId,
+			as.Artifact.SequenceClientId,
+		)
+		if err != nil {
+			return gql.CreatedArtifactArtifact{}, err
+		}
+		return response.GetCreateArtifact().GetArtifact(), nil
+	} else {
+		response, err := gql.CreateArtifactWithoutTags(
+			as.Ctx,
+			as.GraphqlClient,
+			as.Artifact.Entity,
+			as.Artifact.Project,
+			as.Artifact.Type,
+			as.Artifact.Name,
+			runId,
+			as.Artifact.Digest,
+			utils.NilIfZero(as.Artifact.Description),
+			aliases,
+			utils.NilIfZero(as.Artifact.Metadata),
+			utils.NilIfZero(as.Artifact.TtlDurationSeconds),
+			utils.NilIfZero(as.HistoryStep),
+			utils.NilIfZero(as.Artifact.DistributedId),
+			as.Artifact.ClientId,
+			as.Artifact.SequenceClientId,
+		)
+		if err != nil {
+			return gql.CreatedArtifactArtifact{}, err
+		}
+		return response.GetCreateArtifact().GetArtifact(), nil
```

Comment:
context: these 2 blocks are near-duplicates of each other as well

(fwiw, if there are better ways to refactor or revise this logic, open to suggestions)

Reply:
I prefer the method that the python side uses where we just use `CreateArtifact`
and in the `CreateArtifact` graphql we just change the string of the query that we call depending on if the version is earlier or later.
Example: https://github.com/wandb/wandb/blob/main/wandb/sdk/internal/internal_api.py#L3598
Just in case people in the future want to add more fields that need to check the versions for their new field.

Reply:
I agree -- I don't love the idea of defining a new, separate `CreateArtifact` query every time we add a new input field.  Would much rather be able to exclude/include input args programmatically

to clarify -- is the suggestion to bypass the graphql-generated code entirely and just construct the query string here?

Reply:
oh i see the concern about the autogenerated query.... not sure what to do for that either @wandb/sdk-team do y'all have any ideas of how to go around this?

Reply:
@estellazx i agree with your thinking, my best idea is to have wrappers around  the auto generated queries. 
Now we can make so it passes optional arguments (there are 2 styles in go to do it) the version check is done in the wrapper and the relevant arguments are used this way we can always add/remove arguments. 

Honestly, this part is not fully designed so open for suggestions as well

Reply:
> my best idea is to have wrappers around the auto generated queries.
Now we can make so it passes optional arguments (there are 2 styles in go to do it) the version check is done in the wrapper and the relevant arguments are used this way we can always add/remove arguments.

@kptkin not 100% sure I follow yet -- is there maybe an example, code snippet, or doc I can look at to better understand what this might look like?  (happy to sync offline too if faster/easier)

Reply:
no example, yep we can sync offline, i think it will require rewriting the graphql client a little bit